research/picai/fl_nnunet/nnunet_client.py

```diff
@@ -0,0 +1,607 @@
+
+    def create_plans(self, config: Config) -> Dict[str, Any]:
+        """
+        Modifies the provided plans file to work with the local client dataset
+
+        Args:
+            config (Config): The config provided by the server. Expects the
+                'nnunet_plans' key with a pickled dictionary as the value
+
+        Returns:
+            Dict[str, Any]: The modified nnunet plans for the client
+        """
+        # Get the nnunet plans specified by the server
+        plans = pickle.loads(config["nnunet_plans"])  # type: ignore
+
+        # Change plans name
+        if self.plans_name is None:
+            self.plans_name = f"FL_Dataset{self.dataset_id:03d}" + "-" + plans["plans_name"]
+        plans["plans_name"] = self.plans_name
+
+        # Change dataset name
+        plans["dataset_name"] = self.dataset_name
+
+        # Change data identifier and ensure batch size is within nnunet limits
+        # =====================================================================
+        if self.data_identifier is None:
+            self.data_identifier = self.plans_name
+
+        # Max batch size for nnunet is 5 percent of dataset
+        num_samples = self.dataset_json["numTraining"]
+        bs_5percent = round(num_samples * 0.05)
+
+        # Iterate through nnunet configs in plans file
+        for c in plans["configurations"].keys():
+            # Change the data identifier
+            if "data_identifier" in plans["configurations"][c].keys():
+                plans["configurations"][c]["data_identifier"] = self.data_identifier + "_" + c
+            # Ensure batch size is at least 2, then at most 5 percent of dataset
+            if "batch_size" in plans["configurations"][c].keys():
+                old_bs = plans["configurations"][c]["batch_size"]
+                new_bs = max(min(old_bs, bs_5percent), 2)
+                plans["configurations"][c]["batch_size"] = new_bs
+
+        # Can't run nnunet preprocessing without saving plans file
+        if not exists(join(nnUNet_preprocessed, self.dataset_name)):
+            os.makedirs(join(nnUNet_preprocessed, self.dataset_name))
+        plans_save_path = join(nnUNet_preprocessed, self.dataset_name, self.plans_name + ".json")
+        save_json(plans, plans_save_path, sort_keys=False)
+
+        return plans
+
+    def maybe_preprocess(self, nnunet_config: str) -> None:
+        """
+        Checks if preprocessed data for current plans exists and if not preprocesses the nnunet_raw dataset
+
+        Args:
+            nnunet_config (str): string specifying the nnunet config
+        """
+        assert self.data_identifier is not None, "Was expecting data identifier to be initialized in self.create_plans"
+
+        # Preprocess data if it's not already there
+        if self.always_preprocess or not exists(self.nnunet_trainer.preprocessed_dataset_folder):
+            default_num_processes = {"2d": 8, "3d_fullres": 4, "3d_lowres": 8}
```

Comment:
Again same issue with the enum variables not allowed to start with integers. Curious to know what you would suggets for naming

Reply:
Ok I might have gone a bit overkill with this. Went down the rabbit hole of creating an entire subclass of Enum that allows members to have multiple attributes. I know what you meant was to just make the dictionary keys the enum type, but i thought og the idea and thought it would be cool and now its done so. If it's overkill we can scrap the class and just use dictionaries for default_num_processes and num_spatial_dimensions

Reply:
I think it's a very cool subclass and I'm good with keeping it. Makes a lot of the enum stuff very clean 🙂